### PR TITLE
bump netlink&cni&golang version and support add  ecmp route

### PR DESCRIPTION
**Pull request by cyclinder**
Support add ECMP default route:

When I add a `sriov-cni` NIC to the pod by using multus( The first NIC in the pod is assigned by `calico` ). I want the default route for the second NIC (net1) to be created correctly, but there is actually only one default route for calico.
```
root@172-17-8-110:/var/lib/cni/sriov# kubectl exec -it sriov-overlay-8475768cfd-r6slw sh
kubectl exec [POD] [COMMAND] is DEPRECATED and will be removed in a future version. Use kubectl exec [POD] -- [COMMAND] instead.
/ # ip r
default via 169.254.1.1 dev eth0
169.254.1.1 dev eth0 scope link
172.17.8.0/21 dev net1 scope link  src 172.17.8.191
```
This should actually have two default routes:
```
default via 169.254.1.1 dev eth0
default via 172.17.8.1 dev net1  # this is the default route for net1
169.254.1.1 dev eth0 scope link
172.17.8.0/21 dev net1 scope link  src 172.17.8.191
```

In the `calico + macvlan` scenario, where two default routes are created.  I checked the source code of `sriov-cni` and found that the version of CNI it references is too old to cause this problem.

[github.com/containernetworking/cni: v0.8.1](https://github.com/containernetworking/plugins/blob/fe60fcddb897079746ec1523fd1837ab05b1e689/pkg/ipam/ipam_linux.go#L112-L117)

[github.com/containernetworking/cni: v1.1.1](https://github.com/containernetworking/plugins/blob/4744ec27b89c083194e7df498de50f03a8a1d3ec/pkg/ipam/ipam_linux.go#L122)

In summary, I solved this problem by upgrading CNI to `v1.1.1`, and I tested it in my local environment, It works well. As an add-on, we also have to upgrade the `netlink` and `golang` versions. 

That's all this PR is doing.

Signed-off-by: cyclinder <qifeng.guo@daocloud.io>

### CODE REVIEW

**Review by e0ne — APPROVED** [review]

**Comment by zeeke** [comment]
hi @cyclinder , can you provide a little more context about this change?

**Comment by cyclinder** [comment]
Hi @zeeke , I have just updated the context of this change, please review it. Thanks!

**Comment by zeeke** [comment]
@cyclinder thanks for explaining.

nit: is the bump to go1.19 strictly needed? if not, can you at least split it in a different commit?

other than that,
LGTM

Thanks

**Comment by cyclinder** [comment]
> @cyclinder thanks for explaining.
> 
> nit: is the bump to go1.19 strictly needed? if not, can you at least split it in a different commit?
> 
> other than that, LGTM
> 
> Thanks

@zeeke Yeah, bumping to go1.19 is strictly needed. and I could raise another PR to bump to go1.19,After waiting for it to merge to master brach, This PR is ready to merged, WDYT?

**Comment by zeeke** [comment]
@Eoghan1232 please, have a look at this.

**Review by Eoghan1232 — DISMISSED** [review]
overall lgtm, once we clean up the pr to be in line with #228 

we can look to merge both

File: go.mod [hunk 1]
```diff
@@ -1,24 +1,28 @@
 module github.com/k8snetworkplumbingwg/sriov-cni
 
-go 1.17
+go 1.19
```
Comment (Eoghan1232) [inline go.mod:3]: you have pr #228 open for this, can you clean up this pr first before we merge ?

File: go.mod [hunk 2]
```diff
@@ -1,24 +1,28 @@
...
```
Comment (Eoghan1232) [inline go.mod:4]: you have pr #228 open for this, can you clean up this pr first before we merge ?

Comment (cyclinder) [reply go.mod:4]: So can #228 it be merged in first?

**Comment by cyclinder** [comment]
Hi @Eoghan1232 Can you take a look ? thanks

**Review by Eoghan1232 — APPROVED** [review]

**Comment by Eoghan1232** [comment]
lgtm! 
@adrianchiris  can you take a look?

File: go.mod [hunk 3]
```diff
@@ -3,22 +3,26 @@ module github.com/k8snetworkplumbingwg/sriov-cni
...
-	github.com/onsi/ginkgo v1.2.1-0.20151202141238-7f8ab55aaf3b
-	github.com/onsi/gomega v0.0.0-20151007035656-2152b45fa28a
-	github.com/stretchr/testify v1.4.0
-	github.com/vishvananda/netlink v1.0.1-0.20190924205540-07ace697bea4
+	github.com/containernetworking/cni v1.1.2
+	github.com/containernetworking/plugins v1.1.1
+	github.com/onsi/ginkgo v1.16.4
+	github.com/onsi/gomega v1.17.0
+	github.com/stretchr/testify v1.5.1
+	github.com/vishvananda/netlink v1.2.0-beta
```
Comment (adrianchiris) [inline go.mod:11]: is netlink version bump needed ? if so do you know what was added/fixed ?

**Review by adrianchiris — COMMENTED** [review]
Greeting !

while i have no objections with the changes introduced. its a bit unclear to me what fixed ecmp route issue you described.

File: go.mod [hunk 4]
```diff
@@ -3,22 +3,26 @@ module github.com/k8snetworkplumbingwg/sriov-cni
...
 require (
-	github.com/containernetworking/cni v0.8.1
-	github.com/containernetworking/plugins v0.8.7
-	github.com/onsi/ginkgo v1.2.1-0.20151202141238-7f8ab55aaf3b
-	github.com/onsi/gomega v0.0.0-20151007035656-2152b45fa28a
-	github.com/stretchr/testify v1.4.0
-	github.com/vishvananda/netlink v1.0.1-0.20190924205540-07ace697bea4
+	github.com/containernetworking/cni v1.1.2
+	github.com/containernetworking/plugins v1.1.1
+	github.com/onsi/ginkgo v1.16.4
```
[the remark this answers is not in the document]
Comment (cyclinder) [reply go.mod:8]: Hi @adrianchiris , Thanks for your reply!

The key change of this PR is to upgrade the version of CNI and CNI plugin in order to be able to add EcmpRoute. Please refer to the description of the PR for the reason.

In addition, netlink and ginkgo/* are sub-dependencies of CNI and CNI plugins, and after executing `go mod tidy`, it is automatically upgraded.

 So we can't upgrade ginkgo/* and netlink directly, they should be consistent with the dependency versions in cni/cni-plugins

File: go.mod (hunk 3 above)
Comment (cyclinder) [reply go.mod:11]: Yes, In the latest release of `netlink`,  they have added the [RouteAddEcmp](https://github.com/vishvananda/netlink/blob/337a587de15022938a9a31d3f9776faeb5c9f8d7/route_linux.go#L678) function, which is the most important part.

Comment (adrianchiris) [reply go.mod]: OK here is what i found:

support for calling `RouteAddEcmp` was added to `containernetworking/plugins` in v1.0.0 in [this commit](https://github.com/containernetworking/plugins/commit/76ef07ebc6dcfd4fa7c71720544449d6f8e6d693)

in sriov-cni code [we call](https://github.com/k8snetworkplumbingwg/sriov-cni/blob/301a1ffce767a446beb7881402ef1bea99bad71b/cmd/sriov/main.go#L138) `ConfigureInterface` from containernetworking plugins ipam package.


so we need to bump containernetworking/plugins verison.

in v1.1.1 an older version of netlink[ is used](https://github.com/containernetworking/plugins/blob/4744ec27b89c083194e7df498de50f03a8a1d3ec/go.mod#L21)
which apparently is good enough, as bumping netlink version here will not affect the netlink version set in plugins go.mod  (we are using go modules to build so each package provides its own dependencies)

still, i do not think netlink version should be bumped as there is no reason to bump it.
am i missing anything ?

File: go.mod (hunk 4 above)
[the remark this answers is not in the document]
Comment (adrianchiris) [reply go.mod:8]: ack thx for clarifying

File: go.mod (hunk 3 above)
Comment (adrianchiris) [reply go.mod]: ah i see latest PR update now uses netlink v1.1.1 and not 1.2.0

**Comment by cyclinder** [comment]
Thanks @adrianchiris 

Do you have plans to publish a new release?

**Comment by adrianchiris** [comment]
ATM, release is on an "on-demand" basis. can you open an issue requesting a new release ?

**Comment by cyclinder** [comment]
Yes! I could open an issue for this, I think it's meaningful.
